File: src/arena_bot/world.py
"""环境感知：跨 Tick 世界记忆。

记忆规则（对应契约）：
- 障碍是永久地形 → 视野外不丢失，直接可长期使用
- 资源是动态的 → 只做"线索"：VISIBLE（当前可见）/ STALE（曾见现未见）/
  HARVESTED（本侧成功采集）。决策永远优先当前 Turn 的 resource_cells
- 敌人 → 最近位置/类型/时间戳，stale 由调用方按 age 判断，绝不当作当前真相
- 记忆不决定结算：服务器状态是唯一权威
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import TYPE_CHECKING

from .core.unit_state import UnitStateTracker

if TYPE_CHECKING:
    from uuid import UUID

    Position = tuple[int, int]


class ResourceState(Enum):
    VISIBLE = "visible"      # 当前 Turn 可见（自然节点或 cargo 堆）
    STALE = "stale"          # 曾见现未见（可能被采/刷新移走）
    HARVESTED = "harvested"  # 本侧成功采集（节点已消耗）


@dataclass
class ResourceMemory:
    cell: "Position"
    state: ResourceState
    first_seen_tick: int
    last_seen_tick: int


@dataclass
class EnemyMemory:
    id: "UUID"
    position: "Position"
    kind: str  # "UNIT" 或 "CORE"
    last_seen_tick: int


class World:
    """跨 Tick 记忆容器。observe(turn) 每次 Turn 更新。"""

    def __init__(self) -> None:
        self.obstacle_memory: set[Position] = set()
        self.resource_memory: dict[Position, ResourceMemory] = {}
        self.enemy_memory: dict[UUID, EnemyMemory] = {}
        self.failed_cells: dict[Position, int] = {}  # HARVEST_FAILED 格 → tick
        self.tick: int = 0
        self.current_resource_cells: frozenset[Position] = frozenset()
        self.current_enemy_count: int = 0
        self.unit_states = UnitStateTracker()  # 单元意图状态机记忆

    def observe(self, tick: int,
                obstacle_cells: frozenset[Position],
                resource_cells: frozenset[Position],
                visible_enemies: tuple,  # UnitView | CoreView
                harvest_failed_cells: set[Position] = frozenset()) -> None:
        """用当前 Turn 的可见数据更新记忆。"""
        self.tick = tick
        self.current_resource_cells = resource_cells

        # 障碍：永久地形，只增不减
        self.obstacle_memory.update(obstacle_cells)

        # 资源：当前可见 → VISIBLE；曾见的 → STALE
        seen = set()
        for cell in resource_cells:
            seen.add(cell)
            old = self.resource_memory.get(cell)
            self.resource_memory[cell] = ResourceMemory(
                cell=cell, state=ResourceState.VISIBLE,
                first_seen_tick=old.first_seen_tick if old else tick,
                last_seen_tick=tick)
        for cell, mem in list(self.resource_memory.items()):
            if cell in seen:
                continue
            if mem.state is ResourceState.HARVESTED:
                # 已采节点：保持记忆但降级 stale（可能刷新回来，仅作线索）
                mem.state = ResourceState.STALE
                mem.last_seen_tick = tick
            elif mem.state is ResourceState.VISIBLE:
                mem.state = ResourceState.STALE

        # 本侧成功采集/失败：更新状态（由策略在处理事件后传入）
        for cell in harvest_failed_cells:
            self.failed_cells[cell] = tick
            mem = self.resource_memory.get(cell)
            if mem is not None and mem.state is ResourceState.VISIBLE:
                # 失败原因可能是被抢（RESOURCE_DEPLETED）→ 降级，避免反复撞
                mem.state = ResourceState.STALE

        # 敌人：可见更新；不可见的保留最后位置（调用方按 age 用）
        self.current_enemy_count = len(visible_enemies)
        for e in visible_enemies:
            self.enemy_memory[e.id] = EnemyMemory(
                id=e.id, position=e.position, kind=e.kind, last_seen_tick=tick)

    def mark_harvested(self, cell: "Position", tick: int) -> None:
        """成功采集后调用：节点已消耗，从线索中移除。"""
        self.resource_memory[cell] = ResourceMemory(
            cell=cell, state=ResourceState.HARVESTED,
            first_seen_tick=self.resource_memory.get(cell, ResourceMemory(
                cell, ResourceState.STALE, tick, tick)).first_seen_tick,
            last_seen_tick=tick)

    # ---- 查询 ----
    def obstacles(self) -> frozenset["Position"]:
        return frozenset(self.obstacle_memory)

    def resource_hints(self, max_age: int = 8, failed_cooldown: int = 4) -> list["Position"]:
        """最近见过的资源线索（VISIBLE 优先，其次近期 STALE）。

        failed_cooldown：HARVEST_FAILED 的格在此窗口内排除（避免反复撞同格）。
        """
        visible = [m.cell for m in self.resource_memory.values()
                   if m.state is ResourceState.VISIBLE]
        recent = [m.cell for m in self.resource_memory.values()
                  if m.state is ResourceState.STALE
                  and self.tick - m.last_seen_tick <= max_age]
        hints = visible + recent
        return [c for c in hints
                if self.tick - self.failed_cells.get(c, -10**9) >= failed_cooldown]

    def enemy_hints(self, max_age: int = 6) -> list[EnemyMemory]:
        """近期可见敌人线索，按最近出现排序。"""
        return sorted(
            (m for m in self.enemy_memory.values()
             if self.tick - m.last_seen_tick <= max_age),
            key=lambda m: m.last_seen_tick, reverse=True)

File: internal/domain/world.go
package domain

import "sort"

// World 记忆常量（与 TS 版 world.ts 一致）。
const (
	// ResourceMemoryTTLTicks 是 stale/harvested 资源记忆的过期 TTL
	// （≈4 个 refill 周期），防"幽灵资源"——记忆中的资源格实际已被
	// 采空/不再 refill。
	ResourceMemoryTTLTicks = 64
	// DefaultResourceMaxAge 是 resourceHints 中 stale 记忆的最大年龄。
	DefaultResourceMaxAge = 8
	// DefaultFailedCooldown 是 resourceHints 中 HARVEST_FAILED 的冷却。
	DefaultFailedCooldown = 4
	// DefaultMovementCooldownTicks 是 movementObstacles 的默认冷却。
	DefaultMovementCooldownTicks = 3
)

// transientMoveFailureReasons 是单位级动态避让的瞬态移动失败原因
// （与 TS 版 TRANSIENT_MOVE_FAILURE_REASONS 一致）。
var transientMoveFailureReasons = map[string]struct{}{
	"MOVE_CONTESTED":            {},
	"MOVE_SWAP_BLOCKED":         {},
	"MOVE_DESTINATION_OCCUPIED": {},
	"CELL_UNIT_LIMIT":           {},
}

// ResourceState 是资源记忆状态（与 TS 版 ResourceState 一致）。
type ResourceState string

const (
	ResourceStateVisible   ResourceState = "visible"
	ResourceStateStale     ResourceState = "stale"
	ResourceStateHarvested ResourceState = "harvested"
)

// WorkerMode 是单位工作模式（与 TS 版 WorkerMode 一致）。
type WorkerMode string

const (
	WorkerModePatrol    WorkerMode = "patrol"
	WorkerModeGoHarvest WorkerMode = "go_harvest"
)

// ResourceMemory 是资源格记忆（State/LastSeenTick 由 Observe 更新）。
type ResourceMemory struct {
	Cell          Position
	State         ResourceState
	FirstSeenTick int
	LastSeenTick  int
}

// EnemyMemory 是敌方实体记忆。
type EnemyMemory struct {
	ID           string
	Position     Position
	Kind         string
	UnitType     *UnitType
	LastSeenTick int
}

// UnitMemory 是单位工作状态记忆（决策层读写，与 TS 版 UnitMemory 一致）。
type UnitMemory struct {
	WorkerMode      WorkerMode
	HarvestTarget   *Position
	PatrolDirection int
	PatrolRing      int
	PatrolStarted   bool
	PatrolReturning bool
	LastTick        int
}

// ResourceSnapshot 是 WorldSnapshot 中的资源记忆条目。
type ResourceSnapshot struct {
	Cell          string
	State         ResourceState
	FirstSeenTick int
	LastSeenTick  int
}

// EnemySnapshot 是 WorldSnapshot 中的敌方记忆条目。
type EnemySnapshot struct {
	ID           string
	Cell         string
	Kind         string
	LastSeenTick int
}

// WorldSnapshot 是 World 记忆的确定性快照（与 TS 版 snapshot 同语义，
// 切片均按稳定序排列）。
type WorldSnapshot struct {
	Tick      int
	Obstacles []string
	Resources []ResourceSnapshot
	Enemies   []EnemySnapshot
	UnitModes map[string]WorkerMode
}

// World 维护跨 tick 的世界记忆：障碍、资源、敌方实体，以及单位级移动
// 失败冷却（与 TS 版 world.ts World 同语义）：
//   - 障碍永久记忆（直到世界重置）；
//   - 资源记忆状态机 visible → stale/harvested → TTL 过期删除；
//   - 敌方实体按 ID 覆盖更新（位置/类型随新观察刷新）；
//   - tick 回退（服务器世界重置/异常）触发全清并计数。
type World struct {
	tick             int
	obstacleMemory   Set[string]
	resourceMemory   map[string]*ResourceMemory
	enemyMemory      map[string]*EnemyMemory
	failedCells      map[string]int
	unitMoveFailures map[string]map[string]int
	unitMemories     map[string]*UnitMemory

	// WorldResetCount 是世界重置计数（tick 回退检测触发；决策层遥测/测试可读）。
	WorldResetCount int
	// LastWorldResetTick 是最近一次世界重置发生时的 tick（从未重置 = nil）。
	LastWorldResetTick *int
}

// NewWorld 构造空 World。
func NewWorld() *World {
	return &World{
		obstacleMemory:   make(Set[string]),
		resourceMemory:   make(map[string]*ResourceMemory),
		enemyMemory:      make(map[string]*EnemyMemory),
		failedCells:      make(map[string]int),
		unitMoveFailures: make(map[string]map[string]int),
		unitMemories:     make(map[string]*UnitMemory),
	}
}

// Observe 吸收一个规范化 TickState，更新跨 tick 记忆
// （与 TS 版 World.observe 同语义）。
func (w *World) Observe(state *TickState) {
	if w.tick > state.Tick {
		w.reset(state.Tick)
	}
	w.tick = state.Tick

	for cell := range state.ObstacleCells {
		w.obstacleMemory.Add(cell)
	}

	visibleResources := state.ResourceCells.Clone()
	for cell := range visibleResources {
		previous := w.resourceMemory[cell]
		firstSeen := state.Tick
		if previous != nil {
			firstSeen = previous.FirstSeenTick
		}
		position, err := ParseCellKey(cell)
		if err != nil {
			continue
		}
		w.resourceMemory[cell] = &ResourceMemory{
			Cell:          position,
			State:         ResourceStateVisible,
			FirstSeenTick: firstSeen,
			LastSeenTick:  state.Tick,
		}
	}
	for key, memory := range w.resourceMemory {
		if visibleResources.Contains(key) {
			continue
		}
		if memory.State == ResourceStateVisible || memory.State == ResourceStateHarvested {
			memory.State = ResourceStateStale
		}
	}

	for i := range state.Events {
		event := &state.Events[i]
		if event.Position == nil {
			continue
		}
		cell := CellKey(event.Position[0], event.Position[1])
		switch event.EventType {
		case "HARVEST_FAILED":
			w.failedCells[cell] = state.Tick
			if memory := w.resourceMemory[cell]; memory != nil && memory.State == ResourceStateVisible {
				memory.State = ResourceStateStale
			}
		case "HARVEST_SUCCEEDED":
			previous := w.resourceMemory[cell]
			firstSeen := state.Tick
			if previous != nil {
				firstSeen = previous.FirstSeenTick
			}
			position := *event.Position
			w.resourceMemory[cell] = &ResourceMemory{
				Cell:          position,
				State:         ResourceStateHarvested,
				FirstSeenTick: firstSeen,
				LastSeenTick:  state.Tick,
			}
		case "UNIT_MOVE_FAILED":
			if event.ActorID == nil || event.ReasonCode == nil {
				continue
			}
			if _, transient := transientMoveFailureReasons[*event.ReasonCode]; !transient {
				continue
			}
			failures := w.unitMoveFailures[*event.ActorID]
			if failures == nil {
				failures = make(map[string]int)
				w.unitMoveFailures[*event.ActorID] = failures
			}
			failures[cell] = state.Tick
		}
	}

	for i := range state.VisibleEnemies {
		enemy := &state.VisibleEnemies[i]
		w.enemyMemory[enemy.ID] = &EnemyMemory{
			ID:           enemy.ID,
			Position:     enemy.Position,
			Kind:         enemy.Kind,
			UnitType:     enemy.UnitType,
			LastSeenTick: state.Tick,
		}
	}

	liveUnits := make(Set[string], len(state.Units))
	for _, unit := range state.Units {
		liveUnits.Add(unit.ID)
	}
	for unitID := range w.unitMemories {
		if !liveUnits.Contains(unitID) {
			delete(w.unitMemories, unitID)
		}
	}
	for unitID := range w.unitMoveFailures {
		if !liveUnits.Contains(unitID) {
			delete(w.unitMoveFailures, unitID)
		}
	}

	for key, memory := range w.resourceMemory {
		if memory.State != ResourceStateVisible && state.Tick-memory.LastSeenTick > ResourceMemoryTTLTicks {
			delete(w.resourceMemory, key)
			delete(w.failedCells, key)
		}
	}
}

// UnitMemory 返回单位的持久记忆，不存在时以初始巡逻方向创建
// （与 TS 版 World.unitMemory 同语义；返回指针可直接读写）。
func (w *World) UnitMemory(unitID string, initialPatrolDirection int) *UnitMemory {
	memory := w.unitMemories[unitID]
	if memory == nil {
		memory = &UnitMemory{
			WorkerMode:      WorkerModePatrol,
			PatrolDirection: initialPatrolDirection,
		}
		w.unitMemories[unitID] = memory
	}
	memory.LastTick = w.tick
	return memory
}

// Obstacles 返回障碍记忆与额外障碍的并集副本（与 TS 版 obstacles 同语义；
// extra 可为 nil）。
func (w *World) Obstacles(extra Set[string]) Set[string] {
	result := w.obstacleMemory.Clone()
	for cell := range extra {
		result.Add(cell)
	}
	return result
}

// MovementObstacles 返回单位级动态避让集合：服务端 MOVE_CONTESTED 等失败
// 不代表永久地形障碍，只在冷却内阻止同一 actor 重试同一目的格
// （与 TS 版 movementObstacles 同语义；cooldownTicks <= 0 用默认 3）。
func (w *World) MovementObstacles(unitID string, base Set[string], cooldownTicks int) Set[string] {
	if cooldownTicks <= 0 {
		cooldownTicks = DefaultMovementCooldownTicks
	}
	result := base.Clone()
	failures := w.unitMoveFailures[unitID]
	if failures == nil {
		return result
	}
	for cell, failedAt := range failures {
		if w.tick-failedAt < cooldownTicks {
			result.Add(cell)
		} else {
			delete(failures, cell)
		}
	}
	if len(failures) == 0 {
		delete(w.unitMoveFailures, unitID)
	}
	return result
}

// ResourceHints 返回资源格提示：可见优先，stale 按最近见过排序，最近
// 失败的格在冷却内排除（与 TS 版 resourceHints 同语义；maxAge <= 0 用
// 默认 8，failedCooldown <= 0 用默认 4）。
func (w *World) ResourceHints(maxAge, failedCooldown int) []Position {
	if maxAge <= 0 {
		maxAge = DefaultResourceMaxAge
	}
	if failedCooldown <= 0 {
		failedCooldown = DefaultFailedCooldown
	}
	visible := make([]*ResourceMemory, 0)
	recent := make([]*ResourceMemory, 0)
	for _, memory := range w.resourceMemory {
		failedAt, failed := w.failedCells[CellKey(memory.Cell[0], memory.Cell[1])]
		if failed && w.tick-failedAt < failedCooldown {
			continue
		}
		switch memory.State {
		case ResourceStateVisible:
			visible = append(visible, memory)
		case ResourceStateStale:
			if w.tick-memory.LastSeenTick <= maxAge {
				recent = append(recent, memory)
			}
		}
	}
	compare := func(a, b *ResourceMemory) bool {
		if a.LastSeenTick != b.LastSeenTick {
			return a.LastSeenTick > b.LastSeenTick
		}
		if a.Cell[0] != b.Cell[0] {
			return a.Cell[0] < b.Cell[0]
		}
		return a.Cell[1] < b.Cell[1]
	}
	sort.Slice(visible, func(i, j int) bool { return compare(visible[i], visible[j]) })
	sort.Slice(recent, func(i, j int) bool { return compare(recent[i], recent[j]) })
	hints := make([]Position, 0, len(visible)+len(recent))
	for _, memory := range visible {
		hints = append(hints, memory.Cell)
	}
	for _, memory := range recent {
		hints = append(hints, memory.Cell)
	}
	return hints
}

// EnemyHints 返回最近 maxAge ticks 内见过的敌方实体（按最近见过倒序，
// 同 tick 按 ID 升序；与 TS 版 enemyHints 同语义；maxAge <= 0 用默认 6）。
func (w *World) EnemyHints(maxAge int) []EnemyMemory {
	if maxAge <= 0 {
		maxAge = 6
	}
	hints := make([]EnemyMemory, 0)
	for _, memory := range w.enemyMemory {
		if w.tick-memory.LastSeenTick <= maxAge {
			hints = append(hints, *memory)
		}
	}
	sort.Slice(hints, func(i, j int) bool {
		if hints[i].LastSeenTick != hints[j].LastSeenTick {
			return hints[i].LastSeenTick > hints[j].LastSeenTick
		}
		return hints[i].ID < hints[j].ID
	})
	return hints
}

// Snapshot 返回世界记忆的确定性快照（与 TS 版 snapshot 同语义）。
func (w *World) Snapshot() WorldSnapshot {
	obstacles := make([]string, 0, w.obstacleMemory.Len())
	for cell := range w.obstacleMemory {
		obstacles = append(obstacles, cell)
	}
	sort.Strings(obstacles)

	resources := make([]ResourceSnapshot, 0, len(w.resourceMemory))
	for key, memory := range w.resourceMemory {
		resources = append(resources, ResourceSnapshot{
			Cell:          key,
			State:         memory.State,
			FirstSeenTick: memory.FirstSeenTick,
			LastSeenTick:  memory.LastSeenTick,
		})
	}
	sort.Slice(resources, func(i, j int) bool { return resources[i].Cell < resources[j].Cell })

	enemies := make([]EnemySnapshot, 0, len(w.enemyMemory))
	for _, memory := range w.enemyMemory {
		enemies = append(enemies, EnemySnapshot{
			ID:           memory.ID,
			Cell:         CellKey(memory.Position[0], memory.Position[1]),
			Kind:         memory.Kind,
			LastSeenTick: memory.LastSeenTick,
		})
	}
	sort.Slice(enemies, func(i, j int) bool { return enemies[i].ID < enemies[j].ID })

	unitModes := make(map[string]WorkerMode, len(w.unitMemories))
	for id, memory := range w.unitMemories {
		unitModes[id] = memory.WorkerMode
	}

	return WorldSnapshot{
		Tick:      w.tick,
		Obstacles: obstacles,
		Resources: resources,
		Enemies:   enemies,
		UnitModes: unitModes,
	}
}

// reset 是世界重置（tick 回退）：全清本地记忆并计数
// （与 TS 版 observe 内世界重置检测同语义）。
func (w *World) reset(tick int) {
	w.obstacleMemory = make(Set[string])
	w.resourceMemory = make(map[string]*ResourceMemory)
	w.enemyMemory = make(map[string]*EnemyMemory)
	w.failedCells = make(map[string]int)
	w.unitMoveFailures = make(map[string]map[string]int)
	w.unitMemories = make(map[string]*UnitMemory)
	w.WorldResetCount++
	tickCopy := tick
	w.LastWorldResetTick = &tickCopy
}
